README
WHAT IT IS
----------

 musicxml2midi provides an element that converts MusicXML in to MIDI format
suitable for direct synthesis via the wildmidi or timidity elements.


HOW TO USE IT
-------------

 To synthesize MusicXML:
 
  gst-launch filesrc location=song.xml ! musicxml2midi ! wildmidi ! audioconvert ! autoaudiosink

 To convert a MusicXML file in to a MIDI file:

  gst-launch filesrc location=song.xml ! musicxml2midi ! filesink location=song.mid



debian/changelog
gst-musicxml2midi (0.2-0) lucid; urgency=low

  * Give correct caps (application/xml)
  * Add support for auto-plugging

 -- Mike Sheldon <mike@mikeasoft.com>  Mon, 16 Aug 2010 15:25:02 +0100

gst-musicxml2midi (0.1-0) lucid; urgency=low

  * Initial release

 -- Mike Sheldon <mike@mikeasoft.com>  Fri, 19 Mar 2010 12:31:57 +0000


autogen.sh
#!/bin/sh
# you can either set the environment variables AUTOCONF and AUTOMAKE
# to the right versions, or leave them unset and get the RedHat 7.3 defaults

DIE=0
package=gst-musicxml2midi
srcfile=src/main.c

# autogen.sh helper functions (copied from GStreamer's common/ CVS module)
if test ! -f ./gst-autogen.sh;
then
  echo There is something wrong with your source tree.
  echo You are either missing ./gst-autogen.sh or not
  echo running autogen.sh from the top-level source
  echo directory.
  exit 1
fi
. ./gst-autogen.sh

CONFIGURE_DEF_OPT='--enable-maintainer-mode --enable-debug'

autogen_options $@

echo -n "+ check for build tools"
if test ! -z "$NOCHECK"; then echo " skipped"; else  echo; fi
version_check "autoconf" "$AUTOCONF autoconf autoconf259 autoconf257 autoconf-2.54 autoconf-2.53 autoconf-2.52" \
              "ftp://ftp.gnu.org/pub/gnu/autoconf/" 2 52 || DIE=1
version_check "automake" "$AUTOMAKE automake automake-1.9 automake19 automake-1.7 automake-1.6 automake-1.5" \
              "ftp://ftp.gnu.org/pub/gnu/automake/" 1 7 || DIE=1
###version_check "autopoint" "autopoint" \
###              "ftp://ftp.gnu.org/pub/gnu/gettext/" 0 11 5 || DIE=1
version_check "libtoolize" "$LIBTOOLIZE libtoolize glibtoolize" \
              "ftp://ftp.gnu.org/pub/gnu/libtool/" 1 5 0 || DIE=1
version_check "pkg-config" "" \
              "http://www.freedesktop.org/software/pkgconfig" 0 8 0 || DIE=1

die_check $DIE

autoconf_2_52d_check || DIE=1
aclocal_check || DIE=1
autoheader_check || DIE=1

die_check $DIE

# if no arguments specified then this will be printed
if test -z "$*"; then
  echo "+ checking for autogen.sh options"
  echo "  This autogen script will automatically run ./configure as:"
  echo "  ./configure $CONFIGURE_DEF_OPT"
  echo "  To pass any additional options, please specify them on the $0"
  echo "  command line."
fi

tool_run "$aclocal" "-I m4/ $ACLOCAL_FLAGS"
tool_run "$libtoolize" "--copy --force"
tool_run "$autoheader"
tool_run "$autoconf"
tool_run "$automake" "-a -c"

# if enable exists, add an -enable option for each of the lines in that file
if test -f enable; then
  for a in `cat enable`; do
    CONFIGURE_FILE_OPT="--enable-$a"
  done
fi

# if disable exists, add an -disable option for each of the lines in that file
if test -f disable; then
  for a in `cat disable`; do
    CONFIGURE_FILE_OPT="$CONFIGURE_FILE_OPT --disable-$a"
  done
fi

test -n "$NOCONFIGURE" && {
  echo "+ skipping configure stage for package $package, as requested."
  echo "+ autogen.sh done."
  exit 0
}

echo "+ running configure ... "
test ! -z "$CONFIGURE_DEF_OPT" && echo "  ./configure default flags: $CONFIGURE_DEF_OPT"
test ! -z "$CONFIGURE_EXT_OPT" && echo "  ./configure external flags: $CONFIGURE_EXT_OPT"
test ! -z "$CONFIGURE_FILE_OPT" && echo "  ./configure enable/disable flags: $CONFIGURE_FILE_OPT"
echo

./configure $CONFIGURE_DEF_OPT $CONFIGURE_EXT_OPT $CONFIGURE_FILE_OPT || {
        echo "  configure failed"
        exit 1
}

echo "Now type 'make' to compile $package."


src/gstmusicxml2midi.h
/*
 * GStreamer
 * Copyright (C) 2005 Thomas Vander Stichele <thomas@apestaart.org>
 * Copyright (C) 2005 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright (C) 2009 Michael Sheldon <mike@mikeasoft.com>
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef __GST_MUSICXML2MIDI_H__
#define __GST_MUSICXML2MIDI_H__

#include <gst/gst.h>
#include <libxml/parser.h>
#include <libxml/tree.h>

G_BEGIN_DECLS

/* #defines don't like whitespacey bits */
#define GST_TYPE_MUSICXML2MIDIPARSER \
  (gst_musicxml2midi_get_type())
#define GST_MUSICXML2MIDIPARSER(obj) \
  (G_TYPE_CHECK_INSTANCE_CAST((obj),GST_TYPE_MUSICXML2MIDIPARSER,GstMusicXml2MidiParser))
#define GST_MUSICXML2MIDI_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_CAST((klass),GST_TYPE_MUSICXML2MIDIPARSER,GstMusicXml2MidiParserClass))
#define GST_IS_MUSICXML2MIDIPARSER(obj) \
  (G_TYPE_CHECK_INSTANCE_TYPE((obj),GST_TYPE_MUSICXML2MIDIPARSER))
#define GST_IS_MUSICXML2MIDI_CLASS(klass) \
  (G_TYPE_CHECK_CLASS_TYPE((klass),GST_TYPE_MUSICXML2MIDIPARSER))

typedef struct _GstMusicXml2MidiParser      GstMusicXml2MidiParser;
typedef struct _GstMusicXml2MidiParserClass GstMusicXml2MidiParserClass;
typedef struct _Track                 Track;

struct _GstMusicXml2MidiParser
{
  GstElement element;

  GstPad *sinkpad, *srcpad;

  Track *first_track;
  int num_tracks;

  xmlParserCtxtPtr ctxt;
};

struct _GstMusicXml2MidiParserClass
{
  GstElementClass parent_class;
};

struct _Track
{
  xmlChar *xml_id;
  guint16 track_id;
  guint8 midi_channel;
  guint8 midi_instrument;
  guint8 volume;
  guint8 divisions;
  guint8 rest;
  Track *next;
};

GType gst_plugin_musicxml2midi_get_type (void);

G_END_DECLS

#endif /* __GST_MUSICXML2MIDI_H__ */


src/gstmusicxml2midi.c
/*
 * GStreamer
 * Copyright (C) 2005 Thomas Vander Stichele <thomas@apestaart.org>
 * Copyright (C) 2005 Ronald S. Bultje <rbultje@ronald.bitfreak.net>
 * Copyright (C) 2009 Michael Sheldon <mike@mikeasoft.com>
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Alternatively, the contents of this file may be used under the
 * GNU Lesser General Public License Version 2.1 (the "LGPL"), in
 * which case the following provisions apply instead of the ones
 * mentioned above:
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

/**
 * SECTION:element-musicxml2midi
 *
 * Converts MusicXML in to MIDI format suitable for either direct synthesis 
 * via a MIDI element (such as wildmidi or timidity) or saving as a MIDI file.
 *
 * <refsect2>
 * <title>Example launch line</title>
 * |[
 * gst-launch -v -m filesrc location=test.xml ! musicxml2midi ! wildmidi ! audioconvert ! autoaudiosink
 * gst-launch -v -m filesrc location=test.xml ! musicxml2midi ! filesink location=test.mid
 * ]|
 * </refsect2>
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <gst/gst.h>
#include <string.h>
#include <math.h>

#include "gstmusicxml2midi.h"

GST_DEBUG_CATEGORY_STATIC (gst_musicxml2midi_debug);
#define GST_CAT_DEFAULT gst_musicxml2midi_debug

#define TIME_DIVISION 384

/* Filter signals and args */
enum
{
  /* FILL ME */
  LAST_SIGNAL
};

enum
{
  PROP_0
};

/* the capabilities of the inputs and outputs.
 *
 * describe the real formats here.
 */
static GstStaticPadTemplate sink_factory = GST_STATIC_PAD_TEMPLATE ("sink",
    GST_PAD_SINK,
    GST_PAD_ALWAYS,
    GST_STATIC_CAPS ("application/xml")
    );

static GstStaticPadTemplate src_factory = GST_STATIC_PAD_TEMPLATE ("src",
    GST_PAD_SRC,
    GST_PAD_ALWAYS,
    GST_STATIC_CAPS ("audio/midi")
    );

GST_BOILERPLATE (GstMusicXml2Midi, gst_musicxml2midi, GstElement,
    GST_TYPE_ELEMENT);

static void gst_musicxml2midi_set_property (GObject * object, guint prop_id,
    const GValue * value, GParamSpec * pspec);
static void gst_musicxml2midi_get_property (GObject * object, guint prop_id,
    GValue * value, GParamSpec * pspec);

static gboolean gst_musicxml2midi_set_caps (GstPad * pad, GstCaps * caps);
static GstFlowReturn gst_musicxml2midi_chain (GstPad * pad, GstBuffer * buf);
static gboolean gst_musicxml2midi_sink_event (GstPad * pad, GstEvent * event);
static Track *get_track_by_part (GstMusicXml2Midi * filter, xmlChar * part_id);
static GstBuffer *process_element(GstMusicXml2Midi * filter, xmlNode * node);
static GstBuffer *process_partlist(GstMusicXml2Midi * filter, xmlNode * node);
static GstBuffer *process_part(GstMusicXml2Midi * filter, xmlNode * node);
static void process_score_part(GstMusicXml2Midi * filter, xmlNode * node);
static GstBuffer *process_attributes(GstMusicXml2Midi * filter, xmlNode * node, Track *t);
static GstBuffer *process_time(GstMusicXml2Midi * filter, xmlNode * node);
static GstBuffer *process_key(GstMusicXml2Midi * filter, xmlNode * node);
static GstBuffer *process_note(GstMusicXml2Midi * filter, xmlNode * node, Track * track);
static GstBuffer *get_vlv(guint32 val);


/* GObject vmethod implementations */

static void
gst_musicxml2midi_base_init (gpointer gclass)
{
  GstElementClass *element_class = GST_ELEMENT_CLASS (gclass);

  gst_element_class_set_details_simple(element_class,
    "MusicXml2Midi",
    "Parser",
    "Converts musicxml to midi format allowing playback via a standard midi synthesizer",
    "Michael Sheldon <mike@mikeasoft.com>");

  gst_element_class_add_pad_template (element_class,
      gst_static_pad_template_get (&src_factory));
  gst_element_class_add_pad_template (element_class,
      gst_static_pad_template_get (&sink_factory));
}

/* initialize the musicxml2midi's class */
static void
gst_musicxml2midi_class_init (GstMusicXml2MidiClass * klass)
{
  GObjectClass *gobject_class;
  GstElementClass *gstelement_class;

  gobject_class = (GObjectClass *) klass;
  gstelement_class = (GstElementClass *) klass;

  gobject_class->set_property = gst_musicxml2midi_set_property;
  gobject_class->get_property = gst_musicxml2midi_get_property;
}

/* initialize the new element
 * instantiate pads and add them to element
 * set pad calback functions
 * initialize instance structure
 */
static void
gst_musicxml2midi_init (GstMusicXml2Midi * filter,
    GstMusicXml2MidiClass * gclass)
{
  filter->sinkpad = gst_pad_new_from_static_template (&sink_factory, "sink");
  gst_pad_set_setcaps_function (filter->sinkpad,
                                GST_DEBUG_FUNCPTR(gst_musicxml2midi_set_caps));
  gst_pad_set_event_function (filter->sinkpad,
                                GST_DEBUG_FUNCPTR (gst_musicxml2midi_sink_event));
  gst_pad_set_chain_function (filter->sinkpad,
                              GST_DEBUG_FUNCPTR(gst_musicxml2midi_chain));

  filter->srcpad = gst_pad_new_from_static_template (&src_factory, "src");

  gst_element_add_pad (GST_ELEMENT (filter), filter->sinkpad);
  gst_element_add_pad (GST_ELEMENT (filter), filter->srcpad);
  filter->ctxt = xmlCreatePushParserCtxt(NULL, NULL, NULL, 0, NULL);
  filter->first_track = NULL;
  filter->num_tracks = 0;
}

static void
gst_musicxml2midi_set_property (GObject * object, guint prop_id,
    const GValue * value, GParamSpec * pspec)
{
//  GstMusicXml2Midi *filter = GST_MUSICXML2MIDI (object);

  switch (prop_id) {
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
      break;
  }
}

static void
gst_musicxml2midi_get_property (GObject * object, guint prop_id,
    GValue * value, GParamSpec * pspec)
{
//  GstMusicXml2Midi *filter = GST_MUSICXML2MIDI (object);

  switch (prop_id) {
    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
      break;
  }
}

/* GstElement vmethod implementations */

/* this function handles the link with other elements */
static gboolean
gst_musicxml2midi_set_caps (GstPad * pad, GstCaps * caps)
{
  GstMusicXml2Midi *filter;
  GstPad *otherpad;

  filter = GST_MUSICXML2MIDI (gst_pad_get_parent (pad));
  otherpad = (pad == filter->srcpad) ? filter->sinkpad : filter->srcpad;
  gst_object_unref (filter);

  return gst_pad_set_caps (otherpad, caps);
}

/* Convert music xml to MIDI and push it to the src pad */
static GstBuffer *
process_element(GstMusicXml2Midi * filter, xmlNode * node)
{
  xmlNode *cur_node = NULL;
  GstBuffer *buf = NULL;
  GstBuffer *elem_buf;

  for (cur_node = node; cur_node; cur_node = cur_node->next) {
    elem_buf = NULL;
    if (cur_node->type == XML_ELEMENT_NODE) {
      if (xmlStrEqual(cur_node->name, (xmlChar *) "part")) {
        elem_buf = process_part(filter, cur_node);
      } else if (xmlStrEqual(cur_node->name, (xmlChar *) "part-list")) {
        elem_buf = process_partlist(filter, cur_node);
      } else {
        elem_buf = process_element(filter, cur_node->children);
      }
    }
    if (buf == NULL) {
      buf = elem_buf;
    } else if (elem_buf != NULL) {
      buf = gst_buffer_join(buf, elem_buf);
    }
  }

  return buf;
}


static GstBuffer *
process_partlist(GstMusicXml2Midi * filter, xmlNode * node)
{
  xmlNode *child_node = node->children;
  int num_tracks = 0;
  GstBuffer *buf = gst_buffer_new_and_alloc(14);
  char *data = (char *) GST_BUFFER_DATA(buf);
  guint16 *data16 = (guint16 *) GST_BUFFER_DATA(buf);

  while (child_node != NULL) {
    if (xmlStrEqual(child_node->name, (xmlChar *) "score-part")) {
      process_score_part(filter, child_node);
      num_tracks++;
    }
    child_node = child_node->next;
  }

  /* Now we know about the tracks we can send the header */
  memset(data, 0, 14);
  data[0] = 'M'; data[1] = 'T'; data[2] = 'h'; data[3] = 'd'; /* MThd - MIDI File Header */
  data[7] = 6; /* Chunk size */
  data[9] = 1; /* Format */
  data[11] = num_tracks;
  data16[6] = g_htons(TIME_DIVISION);

  return buf;
}

static GstBuffer *
process_part(GstMusicXml2Midi * filter, xmlNode * node)
{
  xmlNode *child_node = node->children;
  xmlNode *measure_node;
  xmlChar *part_id = xmlGetProp(node, (xmlChar *) "id");
  GstBuffer *buf = gst_buffer_new_and_alloc(8);
  char *data = (char *) GST_BUFFER_DATA(buf);
  GstBuffer *note_buf = NULL;
  GstBuffer *tmp_buf = NULL;
  GstBuffer *end_buf = gst_buffer_new_and_alloc(4);
  guint8 *end_data = GST_BUFFER_DATA(end_buf);

  memset(data, 0, 8);
  data[0] = 'M'; data[1] = 'T'; data[2] = 'r'; data[3] = 'k'; /* MTrk - MIDI Track Header */
  end_data[0] = 0x00; end_data[1] = 0xff; end_data[2] = 0x2f; end_data[3] = 0x00; /* End of track event */

  Track *t = get_track_by_part(filter, part_id);
  if (t == NULL) {
    GST_WARNING("No score-part associated with this part. This part will not be heard.");
    return buf;
  }

  while (child_node != NULL) {
    if (xmlStrEqual(child_node->name, (xmlChar *) "measure")) {
        measure_node = child_node->children;
        while (measure_node != NULL) {
          tmp_buf = NULL;
          if (xmlStrEqual(measure_node->name, (xmlChar *) "attributes")) {
            tmp_buf = process_attributes(filter, measure_node, t);
            if (tmp_buf == NULL) {
              measure_node = measure_node->next;
              continue;
            }
            /* Set patch after attributes */
            GstBuffer *patch_buf = gst_buffer_new_and_alloc(3);
            guint8 *patch_data = GST_BUFFER_DATA(patch_buf);
            patch_data[0] = 0; /* Delta time */
            patch_data[1] = 0xc0 | t->midi_channel; /* Set patch on channel */
            patch_data[2] = t->midi_instrument;
            tmp_buf = gst_buffer_join(tmp_buf, patch_buf);
          } else if (xmlStrEqual(measure_node->name, (xmlChar *) "note")) {
            tmp_buf = process_note(filter, measure_node, t);
          }
          
          if (note_buf == NULL) {
            note_buf = tmp_buf;
          } else if (tmp_buf != NULL) {
            note_buf = gst_buffer_join(note_buf, tmp_buf);
          }
          measure_node = measure_node->next;
        }
    }
    child_node = child_node->next;
  }

  note_buf = gst_buffer_join(note_buf, end_buf);

  guint32 *header = (guint32 *) GST_BUFFER_DATA(buf);
  header[1] = g_htonl(GST_BUFFER_SIZE(note_buf));

  buf = gst_buffer_join(buf, note_buf);

  return buf;
}


static Track *
get_track_by_part(GstMusicXml2Midi * filter, xmlChar * part_id) {
  Track *t = filter->first_track;
  while(t != NULL && !xmlStrEqual(t->xml_id, part_id)) {
    t = t->next;
  }
  return t;
}


static void
process_score_part(GstMusicXml2Midi * filter, xmlNode * node)
{
  Track *t = malloc(sizeof(Track));
  Track *n = filter->first_track;
  xmlNode *child_node = node->children;
  xmlNode *midi_child;
  t->track_id = filter->num_tracks;
  t->volume = 127;
  t->midi_channel = 0;
  t->midi_instrument = 0;
  t->divisions = 1;
  t->rest = 0;
  filter->num_tracks++;
  t->xml_id = xmlGetProp(node, (xmlChar *) "id");
  t->next = NULL;

  while (child_node != NULL) {
    if (xmlStrEqual(child_node->name, (xmlChar *) "midi-instrument")) {
      midi_child = child_node->children;
      while (midi_child != NULL) {
        if (xmlStrEqual(midi_child->name, (xmlChar *) "midi-channel")) {
          t->midi_channel = atoi((char *) xmlNodeListGetString(filter->ctxt->myDoc, midi_child->xmlChildrenNode, 1));
        } else if (xmlStrEqual(midi_child->name, (xmlChar *) "midi-program")) {
          t->midi_instrument = atoi((char *) xmlNodeListGetString(filter->ctxt->myDoc, midi_child->xmlChildrenNode, 1));
        }
        midi_child = midi_child->next;
      }
    }
    child_node = child_node->next;
  }

  if (filter->first_track == NULL) {
    filter->first_track = t;
  } else {
    while(n->next != NULL) {
      n = n->next;
    }
    n->next = t;
  }

}


static GstBuffer *
process_attributes(GstMusicXml2Midi * filter, xmlNode * node, Track * t)
{
  xmlNode *child_node = node->children;
  GstBuffer *buf = NULL;
  GstBuffer *tmp_buf;

  while (child_node != NULL) {
    tmp_buf = NULL;
            
    if (xmlStrEqual(child_node->name, (xmlChar *) "time")) {
      tmp_buf = process_time(filter, child_node);
    } else if (xmlStrEqual(child_node->name, (xmlChar *) "key")) {
      tmp_buf = process_key(filter, child_node);
    } else if (xmlStrEqual(child_node->name, (xmlChar *) "divisions")) {
      t->divisions = atoi((char *) xmlNodeListGetString(filter->ctxt->myDoc, child_node->xmlChildrenNode, 1));
    }

    if (buf == NULL) {
      buf = tmp_buf;
    } else if (tmp_buf != NULL) {
      buf = gst_buffer_join(buf, tmp_buf);
    }

    child_node = child_node->next;
  }

  return buf;
}


static GstBuffer *
process_time(GstMusicXml2Midi * filter, xmlNode * node)
{
  xmlNode *child_node = node->children;
  GstBuffer *buf = gst_buffer_new_and_alloc(8);
  guint8 *data = GST_BUFFER_DATA(buf);
  guint8 beats = 0;
  guint8 beat_type = 0;

  while (child_node != NULL) {
    if (xmlStrEqual(child_node->name, (xmlChar *) "beats")) {
      beats = atoi((char *) xmlNodeListGetString(filter->ctxt->myDoc, child_node->xmlChildrenNode, 1));
    } else if (xmlStrEqual(child_node->name, (xmlChar *) "beat-type")) {
      beat_type = atoi((char *) xmlNodeListGetString(filter->ctxt->myDoc, child_node->xmlChildrenNode, 1));
    }
    child_node = child_node->next;
  }

  if(beats != 0 && beat_type != 0) {
    data[0] = 0x00; /* Delta time */
    data[1] = 0xff; /* Meta event */
    data[2] = 0x58; /* Set time signature */
    data[3] = 4; /* Event data length */
    data[4] = beats;
    data[5] = (guint8) sqrt(beat_type);
    data[6] = 24; /* Metronome */
    data[7] = 8; /* 32nds */
    return buf;
  } else {
    return NULL;
  }
}


static GstBuffer *
process_key(GstMusicXml2Midi * filter, xmlNode * node)
{
  xmlNode *child_node = node->children;
  GstBuffer *buf = gst_buffer_new_and_alloc(6);
  guint8 *data = GST_BUFFER_DATA(buf);
  guint8 fifths = 0;

  while (child_node != NULL) {
    if (xmlStrEqual(child_node->name, (xmlChar *) "fifths")) {
      fifths = atoi((char *) xmlNodeListGetString(filter->ctxt->myDoc, child_node->xmlChildrenNode, 1));
    }
    child_node = child_node->next;
  }

  data[0] = 0x00; /* Delta time */
  data[1] = 0xff; /* Meta event */
  data[2] = 0x59; /* Set key signature */
  data[3] = 2; /* Event data length */
  data[4] = fifths;
  data[5] = 0; /* Scale */

  return buf;
}


static GstBuffer *
process_note(GstMusicXml2Midi * filter, xmlNode * node, Track * track)
{
  xmlNode *child_node = node->children;
  xmlNode *pitch_child;
  GstBuffer *buf = gst_buffer_new_and_alloc(3);
  GstBuffer *off_buf = gst_buffer_new_and_alloc(3);
  guint8 *data = GST_BUFFER_DATA(buf);
  guint8 *off_data = GST_BUFFER_DATA(off_buf);
  guint8 duration = 0, pitch = 0, step = 0, octave = 0;
  gint8 alter = 0;
  gboolean rest = FALSE;

  while (child_node != NULL) {
    if (xmlStrEqual(child_node->name, (xmlChar *) "duration")) {
      duration = atoi((char *) xmlNodeListGetString(filter->ctxt->myDoc, child_node->xmlChildrenNode, 1));
    } else if (xmlStrEqual(child_node->name, (xmlChar *) "rest")) {
      rest = TRUE;
    } else if (xmlStrEqual(child_node->name, (xmlChar *) "pitch")) {
      pitch_child = child_node->children;
      while (pitch_child != NULL) {
        if (xmlStrEqual(pitch_child->name, (xmlChar *) "step")) {
          step = (guint8) xmlNodeListGetString(filter->ctxt->myDoc, pitch_child->xmlChildrenNode, 1)[0]; 
          if (step < 72) {
            /* ASCII offset to capital C */
            step -= 67;
          } else {
            /* ASCII offset to lower case c */
            step -= 99;
          }
          step = step * 2;
          if (step > 5) { /* E */
            step--;
          }
        } else if (xmlStrEqual(pitch_child->name, (xmlChar *) "octave")) {
          octave = atoi((char *) xmlNodeListGetString(filter->ctxt->myDoc, pitch_child->xmlChildrenNode, 1));
        } else if (xmlStrEqual(pitch_child->name, (xmlChar *) "alter")) {
          alter = atoi((char *) xmlNodeListGetString(filter->ctxt->myDoc, pitch_child->xmlChildrenNode, 1));
        }
        pitch_child = pitch_child->next;
      }
      pitch = (12 * (octave + 1)) + step + alter; /* Convert to MIDI note numbers (C4 == 60) */
    } 
    child_node = child_node->next;
  }

  if (rest) {
    track->rest += duration;  
  } else {
    /* Note on */
    data[0] = 0x90 | track->midi_channel;
    data[1] = pitch;
    data[2] = track->volume;
    buf = gst_buffer_join(get_vlv(track->rest * TIME_DIVISION / track->divisions), buf);
    track->rest = 0;

    /* Note off */
    off_data[0] = 0x80 | track->midi_channel;
    off_data[1] = pitch;
    off_data[2] = 0;
    off_buf = gst_buffer_join(get_vlv(duration * TIME_DIVISION / track->divisions), off_buf);
    buf = gst_buffer_join(buf, off_buf);
  }

  return buf;
}


/* Return a buffer containing a variable length value (vlv)
 * suitable for use as delta times */
static GstBuffer * 
get_vlv (guint32 val)
{
  // Calculate how many bytes we need
  int len = 1;
  while ( ((val >> 7 * len) & 0x7f) != 0 ) {
    len++;
  }

  GstBuffer *buf = gst_buffer_new_and_alloc(len);
  guint8 *data = GST_BUFFER_DATA(buf);
  int i;
  
  for(i = len - 1; i > -1; i--) {
    if (i != 0) {
      data[(len - 1) - i] = 0x80 | ((val >> 7 * i) & 0x7f);
    } else {
      data[(len - 1) - i] = val & 0x7f;
    }
  }

  return buf;
} 


static gboolean
gst_musicxml2midi_sink_event (GstPad * pad, GstEvent * event)
{
  if (GST_EVENT_TYPE(event) == GST_EVENT_EOS) {
    GstMusicXml2Midi *filter = GST_MUSICXML2MIDI (gst_pad_get_parent (pad));
    xmlNode *root = xmlDocGetRootElement(filter->ctxt->myDoc);
    GstBuffer *buf = process_element(filter, root);
    gst_pad_push (filter->srcpad, buf);
  } 
  return gst_pad_event_default (pad, event);
}


/* chain function
 * this function does the actual processing
 */
static GstFlowReturn
gst_musicxml2midi_chain (GstPad * pad, GstBuffer * buf)
{
  GstMusicXml2Midi *filter;
  xmlNode *root_element = NULL;

  filter = GST_MUSICXML2MIDI (gst_pad_get_parent (pad));

  xmlParseChunk(filter->ctxt, (char *) buf->data, buf->size, 0);
  root_element = xmlDocGetRootElement(filter->ctxt->myDoc);

  return GST_FLOW_OK;

//  return gst_pad_push (filter->srcpad, buf);
}


/* entry point to initialize the plug-in
 * initialize the plug-in itself
 * register the element factories and other features
 */
static gboolean
musicxml2midi_init (GstPlugin * musicxml2midi)
{
  /* debug category for filtering log messages
   *
   */
  GST_DEBUG_CATEGORY_INIT (gst_musicxml2midi_debug, "musicxml2midi",
      0, " musicxml2midi");

  return gst_element_register (musicxml2midi, "musicxml2midi", GST_RANK_MARGINAL,
      GST_TYPE_MUSICXML2MIDI);
}

/* gstreamer looks for this structure to register musicxml2midis
 */
GST_PLUGIN_DEFINE (
    GST_VERSION_MAJOR,
    GST_VERSION_MINOR,
    "musicxml2midi",
    "Converts musicxml to midi format allowing playback via a standard midi synthesizer",
    musicxml2midi_init,
    VERSION,
    "LGPL",
    "Michael Sheldon <mike@mikeasoft.com>",
    "http://github.com/Elleo/gst-musicxml2midi"
)
